### DISCUSSION TITLE
CAM and ScoreCAM do not match. ScoreCAM yields visually similar activation maps for binary classification.

### DISCUSSION THREAD

**Discussion by CHILLQQ** [frgfm/torch-cam#213] [Q&A] [answered]
Hello,

I have trained a small neural net for binary classification and I am trying to get the class activation maps using CAM and ScoreCAM methods. The architecture of the neural net has global average pooling layer followed by an fc layer so that it is suitable for CAM.
I just wanted to make sure that CAM and ScoreCAM yield same activation maps. But this seems not to be the case.
As I only have 2 classes, my activation maps obtained by CAM are exact opposite of each other which makes sense to me. But, when I use ScoreCAM I get visually identical activation maps when using _class_idx_=0 and _class_idx_=1. Do you know why this could be the case? I don't immediately see how ScoreCAM reduces to CAM when the nn architecture is chosen to fit CAM, so it is hard for me to find the inconsistency.

Thank you very much.
Mikhail

**Comment by frgfm** [comment, marked as answer]
Hey @CHILLQQ :wave: 

Thanks for the question! For clarifications:
- what do you mean by "make sure that CAM and ScoreCAM yield same activation maps" ? They are different methods so it's normal that it yields different outputs, and no, ScoreCAM cannot be reduced to CAM (cf. equations in https://frgfm.github.io/torch-cam/methods.html#torchcam.methods.ScoreCAM)
- however about the other problem you're mentioning, I can only see only 2 cases where switching between two  class_idx gets you the same activation map:
    - if there is a NaN & the activation map is uniform
    - if the activations are right on prob=0.5

To be able to help you further, I'd need a minimal runnable snippet so that I can reproduce :sweat_smile: 

Let me if that helped or if you can share a snippet,
Cheers :v:

**Reply by CHILLQQ** [reply]
Thank you for the quick response @frgfm. 

I didn't expect CAM and Score CAM activation maps to be identical, but the have nothing in common at all. 
I tried implementing my own ScoreCAM as well, and I got the same problem of the visually identical heatmaps for the two classes. I am saying this because there I for sure did not see prob of 0.5 nor any NaN values.
I shared a little [collab](https://colab.research.google.com/drive/1EIB4hWmEe4M_KZo14DeifqYP8EfA03b0?usp=sharing) notebook that reproduces my problem.


Thank you for your help!

**Reply by frgfm** [reply]
Especially is you get the same results, I'd advise two things: checking whether
- the output value of class 0 is actually different from 1 (for some reasons perhaps there was an issue during training leading to having similar value). If they're different check by how much (mean val of abs difference)
- check whether, because resizing, the activation maps are actually the same

My guess is that your model is outputting the same. When you said binary classification I assumed 1 output node (so 99% means 1% for the other), but it's actually 2 nodes so it's entirely possible that you have the same value more or so for both :sweat_smile: 

Let me know what you find!

**Reply by CHILLQQ** [reply]
Thank you for your help!
To be honest, I don't really need two output neurons. 
But if I use only one output neuron, does it imply that all heatmaps I get will be for class=1, and to get the heatmap for the class=0, I would need to do _heatmap_class0 = 1 - heatmap_class1_?
I am asking because after using 1 output neuron the CAM heatmaps are the exact opposite of ScoreCAM heatmaps (_heatmap_CAM = 1-heatmap_ScoreCAM_), which doesn't look right.

**Reply by frgfm** [reply]
Yes correct, `1 - cam` should get you the CAM for the other class!

The second part is suspicious though. Could you try to install the [dev version](https://github.com/frgfm/torch-cam#developer-installation) instead and see if you still have the problem?
